File: client/src/model/PythonModule.ts
import PythonFunction from "./PythonFunction";
import PythonClass from "./PythonClass";
import PythonFromImport from "./PythonFromImport";
import PythonImport from "./PythonImport";

export default class PythonModule {

    readonly name: string;
    readonly imports: PythonImport[];
    readonly fromImports: PythonFromImport[];
    readonly classes: PythonClass[];
    readonly functions: PythonFunction[];

    constructor(
        name: string,
        imports: PythonImport[] = [],
        fromImports: PythonFromImport[] = [],
        classes: PythonClass[] = [],
        functions: PythonFunction[] = []
    ) {
        this.name = name;
        this.imports = imports;
        this.fromImports = fromImports;
        this.classes = classes;
        this.functions = functions;
    }

    toString() {
        return `Module "${this.name}"`
    }
}

File: api-editor/desktop/src/main/kotlin/com/larsreimann/apiEditor/features/ast/model/PythonModule.kt
package com.larsreimann.apiEditor.features.ast.model

interface PythonModule : PythonDeclaration {
    override val id: PythonModuleId
    val imports: List<PythonImport>
    val fromImports: List<PythonFromImport>
    val classes: List<PythonClass>
    val functions: List<PythonFunction>

    override fun children() = sequence {
        yieldAll(classes)
        yieldAll(functions)
    }

    fun toMutablePythonModule() = MutablePythonModule(
        id = id,
        imports = imports,
        fromImports = fromImports,
        classes = classes.map { it.toMutablePythonClass() },
        functions = functions.map { it.toMutablePythonFunction() },
    )
}

class MutablePythonModule(
    override val id: PythonModuleId = PythonModuleId(""),
    override var name: String = "",
    imports: List<PythonImport> = emptyList(),
    fromImports: List<PythonFromImport> = emptyList(),
    classes: List<MutablePythonClass> = emptyList(),
    functions: List<MutablePythonFunction> = emptyList(),
) : MutablePythonDeclaration(), PythonModule {

    override val imports = imports.toMutableList()
    override val fromImports = fromImports.toMutableList()
    override val classes = MutableContainmentList(classes)
    override val functions = MutableContainmentList(functions)
}

typealias PythonModuleId = PythonDeclarationId<PythonModule>
